File: C/One_sided/Fencepost/simple_allocate.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100
#define NELEMENTS 10

int main(int argc, char ** argv)
{

  int rank, recv_rank, nproc, left, right, iindex, offset_p;
  int *data_out;
  int data_in[NELEMENTS];
  char outstring[100];
  MPI_Aint size_of_window, offset;
  MPI_Win window;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  offset = 0;
  size_of_window = sizeof(int) * NELEMENTS;

  for (iindex = 0; iindex<NELEMENTS;++iindex){
    data_in[iindex] = iindex + rank;
  }

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  //Allocate memory
  MPI_Alloc_mem(size_of_window, MPI_INFO_NULL, &data_out);

  //Create the window. This is a piece of memory that's available for remote
  //access. In this case, a single 4 byte integer
  MPI_Win_create(data_out, size_of_window, sizeof(int), MPI_INFO_NULL,
      MPI_COMM_WORLD, &window);

  //Use collective synchronization model. After this command any processor
  //can use MPI_Put or MPI_Get on any other processor
  MPI_Win_fence(0,window);

  //Put the result into the first (zeroth) slot
  offset = 0;
  //Actual call to put the data in the remote processor
  MPI_Put(data_in, NELEMENTS, MPI_INT, right, offset, NELEMENTS, MPI_INT,
      window);

  //Call Win_fence again to end the access and exposure epochs
  MPI_Win_fence(0, window);

  offset_p = sprintf(outstring, "Rank %3d got message from rank %3d of ",
      rank, left);

  for (iindex = 0; iindex<NELEMENTS; ++iindex){
    offset_p += sprintf(outstring+offset_p, "%3d ", data_out[iindex]);
  }
  printf("%s\n", outstring);

  MPI_Free_mem(data_out);

  MPI_Finalize();

}

File: Fortran/One_sided/Fencepost/simple_allocate.f90
PROGRAM mpi_fence

  USE mpi
  USE iso_c_binding
  IMPLICIT NONE

  INTEGER, PARAMETER :: n_elements = 10

  INTEGER :: rank, recv_rank, nproc, iindex
  INTEGER :: left, right
  INTEGER :: ierr, window, intsize
  INTEGER(KIND=MPI_ADDRESS_KIND) :: size_of_window, offset
  TYPE(C_PTR) :: c_pointer
  INTEGER, DIMENSION(n_elements) :: data
  INTEGER, DIMENSION(:), POINTER :: f_pointer

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Set up periodic domain
  left = rank - 1
  IF (left < 0) left = nproc - 1
  right = rank + 1
  IF (right > nproc - 1) right = 0

  !In C would just use sizeof()
  !In Fortran2008, could use C_SIZEOF
  !NOTE! This routine returns the size of AN ELEMENT
  !Not the size of an array if you give it an array
  CALL MPI_Sizeof(recv_rank, intsize, ierr)
  !Just using a single int here
  size_of_window = intsize * n_elements

  !Allocate the memory
  CALL MPI_Alloc_mem(size_of_window, MPI_INFO_NULL, c_pointer, ierr)

  !Get Fortran pointer to 
  CALL C_F_POINTER(c_pointer, f_pointer, shape=(/n_elements/))

  !Create the window. This is a piece of memory that's available for remote
  !access. In this case, a single 4 byte integer
  CALL MPI_Win_create(f_pointer, size_of_window, intsize, MPI_INFO_NULL, &
      MPI_COMM_WORLD, window, ierr)

  !Populate data object
  DO iindex = 1, n_elements
    data(iindex) = iindex + rank
  END DO

  !Use collective synchronization model. After this command any processor
  !can use MPI_Put or MPI_Get on any other processor
  CALL MPI_Win_fence(0, window, ierr)

  !Put the result into the first (zeroth) slot
  offset = 0
  !Actual call to put the data in the remote processor
  CALL MPI_Put(data, n_elements, MPI_INTEGER, right, offset, n_elements, &
      MPI_INTEGER, window, ierr)

  !Call Win_fence again to end the access and exposure epochs
  CALL MPI_Win_fence(0, window, ierr)

  PRINT ("(a,i3, a, i3, a, 10i3)"),"Rank ", rank, " got message from rank ", &
      left, " of ", f_pointer

  CALL MPI_Free_mem(f_pointer, ierr)

  CALL MPI_Finalize(ierr)

END PROGRAM mpi_fence
